core/src/main/java/org/infinispan/atomic/impl/AtomicKeySetImpl.java

```diff
@@ -554,4 +557,88 @@ public Externalizable readObject(ObjectInput input) throws IOException, ClassNot
          }
       }
    }
+
+   public static class KeyMarshaller implements MessageMarshaller<AtomicKeySetImpl.Key> {
```

Comment:
Was wondering why we use a manually written marshaller for this? Any protostream bug that had to be worked around?

Please keep in mind that manual marshallers have the potential of being 10x slower than generated marshallers.  IMO, we should deprecate MessageMarshaller in next major protostream vesion.

Reply:
Hmm I can't remember, but I'm hoping that https://github.com/infinispan/infinispan/pull/6969 will get integrated so that we can remove all of the manual marshallers anyway. 

A tentative +1 for removing manual marshallers in future protostream versions. An obvious objection would be if AtomicMaps don't get removed, but another concern is that they might be necessary in some obscure part of our code when converting the GlobalMarshaller to utilise protostream.

Reply:
@anistor I'm guessing it's because it must use `persistenceMarshaller.marshallUserObject`.

Could we write custom marshallers that use the same approach as the generated marshallers for speed? I'm thinking we could keep fields immutable if we read all the fields from the message in a loop, saved them in local variables, and invoked a constructor at the end.

OTOH I would very much like to replace `MessageMarshaller` with something else, I don't like how overloaded `Marshaller` has become in our code base with this PR.

Reply:
@danberindei In terms of this PR, now that AtomicMaps have been removed I don't think we need any manual marshaller implementations.

I do like the idea of being able to write performant marshallers by hand though.

Reply:
For now I would simply like to avoid manually written marshallers if possible. This is not the right place to discuss protostream extensions.